Bottom: 684f57ed391827a478d2c848140dfc2c75cb88c4
Top:    75153e86dee441c11639cc0d142e7aa67c2eeb66
Author: Jason Wang <jasowang@redhat.com>
Date:   2017-02-16 16:51:33 +0800

Refresh of virtio-net-optimize-small

---

diff --git a/drivers/net/virtio_net.c b/drivers/net/virtio_net.c
index 5866271..98a7e93 100644
--- a/drivers/net/virtio_net.c
+++ b/drivers/net/virtio_net.c
@@ -460,7 +460,10 @@ static struct sk_buff *receive_small(struct net_device *dev,
 	}
 	rcu_read_unlock();
 
+	skb = build_skb(buf + vi->hdr_len, GOOD_PACKET_LEN);
 	skb_trim(skb, len);
+	memcpy(skb_vnet_hdr(skb), buf, vi->hdr_len);
+
 	return skb;
 
 err_xdp:
@@ -796,26 +799,26 @@ static unsigned int virtnet_get_headroom(struct virtnet_info *vi)
 static int add_recvbuf_small(struct virtnet_info *vi, struct receive_queue *rq,
 			     gfp_t gfp)
 {
-	int headroom = GOOD_PACKET_LEN + virtnet_get_headroom(vi);
+	const size_t hdr_len = sizeof(struct virtio_net_hdr);
+	struct page_frag *alloc_frag = &rq->alloc_frag;
+	char *buf;
 	unsigned int xdp_headroom = virtnet_get_headroom(vi);
-	struct sk_buff *skb;
-	struct virtio_net_hdr_mrg_rxbuf *hdr;
+	int len = hdr_len + NET_IP_ALIGN + GOOD_PACKET_LEN + xdp_headeroom;
 	int err;
 
-	skb = __netdev_alloc_skb_ip_align(vi->dev, headroom, gfp);
-	if (unlikely(!skb))
+	len = SKB_DATA_ALIGN(len) +
+	      SKB_DATA_ALIGN(sizeof(struct skb_shared_info));
+	if (unlikely(!skb_page_frag_refill(len, alloc_frag, gfp)))
 		return -ENOMEM;
 
-	skb_put(skb, headroom);
-
-	hdr = skb_vnet_hdr(skb);
-	sg_init_table(rq->sg, 2);
-	sg_set_buf(rq->sg, hdr, vi->hdr_len);
-	skb_to_sgvec(skb, rq->sg + 1, xdp_headroom, skb->len - xdp_headroom);
+	buf = (char *)page_address(alloc_frag->page) + alloc_frag->offset;
+	sg_init_table(rq->sg, 1);
+	sg_set_buf(rq->sg + 1, buf + NET_IP_ALIGN + xdp_headeroom,
+		   GOOD_PACKET_LEN);
 
-	err = virtqueue_add_inbuf(rq->vq, rq->sg, 2, skb, gfp);
+	err = virtqueue_add_inbuf(rq->vq, rq->sg, 1, buf, gfp);
 	if (err < 0)
-		dev_kfree_skb(skb);
+		put_page(virt_to_head_page(buf));
 
 	return err;
 }
